core/src/main/java/org/lflang/generator/c/CReactionGenerator.java

```diff
@@ -561,11 +563,13 @@ private static String generateActionVariablesInReaction(
                 + " != NULL && "
                 + tokenPointer
                 + "->value != NULL);",
+            "lf_critical_section_enter(self->base.environment);",
```

Comment:
I think the critical section needs to include the two operations above, setting is_present and has_value.  Also the setting of the value below.